File: datastructures/stack/stackAsArray.js
class Stack {
    constructor() {
        this.top = -1;
        this.maxNum = 100;
        this.arrStack = new Array(this.maxNum);

        this.push = this.push.bind(this);
        this.pop = this.pop.bind(this);
        this.returnTop = this.returnTop.bind(this);
        this.isEmpty = this.isEmpty.bind(this);
        this.printStack = this.printStack.bind(this);
    }

    push(x) {
        if (this.top == (this.maxNum - 1)) {
            var temp = new Array(this.maxNum * 2);
            for (var i = 0; i < this.arrStack.length; i++) {
                temp[i] = this.arrStack[i];
            }
            this.arrStack = temp;
        }
        this.top++;
        this.arrStack[this.top] = x;
    }

    pop() {
        if (this.top == -1) {
            return console.log("The stack is already empty.");
        }
        this.top--;
        this.returnTop();
    }

    returnTop() {
        if (this.top == -1) {
            console.log("The stack is empty.");
        }
        return this.top;
    }

    isEmpty() {
        return (this.top == -1) ?  true : false;
    }

    printStack() {
        if (this.top == - 1) return console.log("Sorry, the stack is empty.");

        var returnString = "The stack is: ";
        for (var i = 0; i <= this.top; i++) {
            returnString += ("[" + this.arrStack[i] + "] ");
        }
        returnString += ".";
        return returnString;
    }
}

console.log('Starting job...');

var myStack = new Stack();
console.log('Is empty? ' + myStack.isEmpty());
console.log('Top? ' + myStack.returnTop());
myStack.pop();
myStack.push(3);
myStack.push(5);
myStack.push(7);
myStack.push(9);
console.log(myStack.printStack());
myStack.pop();
console.log(myStack.printStack());
console.log('Top? ' + myStack.returnTop());
console.log('Is empty? ' + myStack.isEmpty());

console.log('Job ended.');

File: datastructures/stack/stackAsArray.java
class Stack {
    int top = -1;
    int maxNum = 100;
    int[] arrStack;

    public Stack() {
        arrStack = new int[maxNum];
    }

    public void push(int x) {
        if (top == (maxNum - 1)) {
            int[] temp = new int[maxNum * 2];
            for (int i = 0; i < arrStack.length; i++) {
                temp[i] = arrStack[i];
            }
            arrStack = temp;
        }
        top = top + 1;
        arrStack[top] = x;
    }

    public void pop() {
        if (top == -1) {
            System.out.println("Stack is already empty.");
            return;
        }
        top = top - 1;
        top();
    }

    public int top() {
        if (top == -1) {
            System.out.println("Stack is empty.");
            return 0;
        }
        return arrStack[top];
    } 

    public boolean isEmpty() {
        if (top == -1) {
            return true;
        } else {
            return false;
        }
    }

    public void printStack() {
        System.out.print("The stack is: ");
        for (int i = 0; i <= top; i++) {
            System.out.print("[" + arrStack[i] + "] ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Stack myStack = new Stack();
        System.out.println("Empty? " + myStack.isEmpty());
        myStack.pop();
        myStack.top();
        myStack.push(1);
        myStack.push(3);
        myStack.push(5);
        myStack.push(7);
        myStack.printStack();
        myStack.pop();
        myStack.printStack();
    }
}